run_so2sat_experiments.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""Runs the train script with a grid of hyperparameters."""
+import itertools
+import os
+import subprocess
+from multiprocessing import Process, Queue
+
+# list of GPU IDs that we want to use, one job will be started for every ID in the list
+GPUS = [0, 1, 2, 3, 4, 5, 6, 7]
+TEST_MODE = False  # if False then print out the commands to be run, if True then run
+
+# Hyperparameter options
+model_options = ["resnet18", "resnet50"]
+lr_options = [1e-2, 1e-3, 1e-4]
+loss_options = ["ce"]
+weight_options = ["imagenet_only", "random", "imagenet_and_random", "random_rgb"]
+
+
+def do_work(work: "Queue[str]", gpu_idx: int) -> bool:
+    """Process for each ID in GPUS."""
+    while not work.empty():
+        experiment = work.get()
+        experiment = experiment.replace("GPU", str(gpu_idx))
+        print(experiment)
+        if not TEST_MODE:
+            subprocess.call(experiment.split(" "))
+    return True
+
+
+def main() -> None:
+    """Main."""
+    work: Queue[str] = Queue()
+
+    for (model, lr, loss, weights) in itertools.product(
+        model_options,
+        lr_options,
+        loss_options,
+        weight_options,
+    ):
+
+        experiment_name = f"{model}_{lr}_{loss}_{weights.replace('_','-')}_reproduce1"
+
+        output_dir = "output/so2sat_experiments/"
+
+        if not os.path.exists(os.path.join(output_dir, experiment_name)):
+
+            command = (
+                "python train.py"
+                + " config_file=conf/so2sat.yaml"
+                + f" experiment.name={experiment_name}"
+                + f" experiment.module.classification_model={model}"
+                + f" experiment.module.learning_rate={lr}"
+                + f" experiment.module.loss={loss}"
+                + f" experiment.module.weights={weights}"
+                + f" experiment.datamodule.weights={weights}"
+                + f" program.output_dir={output_dir}"
+                + f" program.log_dir={output_dir}/logs"
+                + " program.data_dir=/home/calebrobinson/ssdprivate/data/so2sat"
```

Comment:
I don't think this directory should be hard-coded. Could it be an option to the script? In terms of reproducibility, things like assuming users will have 8 GPUs on their system or using specific paths isn't great. Although you could argue that this script is only used to tune hyperparameters, and users can use whatever they want and run `train.py` directly.

Reply:
I don't think these scripts need to be made pretty, interested users will be able to figure it out

Reply:
This isn't a matter of appearance, it's a matter of reproducibility. Yes, someone might be able to figure it out, but the more difficult it is to reproduce the less we have the right to brag about reproducibility in our own paper.

Reply:
There's always a more complex way to do it, I _could_ add argparse to these, but I don't think it helps anyone "reproduce" our results easier. Changing a constant at the top of the script vs. setting an argparse flag takes the same amount of comprehension (of what the parameters mean) and effort. I don't use argparse here because these scripts are meant to be mostly static (i.e. if I do, the hyperparameter ranges are _not_ arguments), but I'm happy to set it up if you think that is absolutely necessary.